docs/dg/learningBasics.md

```diff
@@ -0,0 +1,416 @@
+        isSinceDateProvided, isUntilDateProvided, numCloningThreads, numAnalysisThreads, formats,
+        shouldIncludeLastModifiedDate, shouldPerformShallowCloning, isAutomaticallyLaunching,
+        isStandaloneConfigIgnored, zoneId, isJsonPrettyPrintingUsed);
+``` 
+</panel>
+
+<panel header="Hint 3">
+
+After the steps in hint 1 and hint 2, the result returned from `ArgsParser.parse(args)` in `RepoSense.java` should already be able to capture the new argument when it is specified in the command. 
+
+The next step is to pass the argument to JSON file writer to notify it of the specified printing mode.
+
+Note that the creation and writing of JSON file is invoked in `ReportGenerator.generateReposReport`, which calls `FileUtil.writeJsonFile` directly to write the `summary.json` file or indirectly (Check `generateIndividualRepoReport` and `generateEmptyRepoReport`) to write the `commits.json` or `authorship.json` files of individual repositories.
+
+Therefore, the task now is to make `FileUtil.writeJsonFile` switch between different printing mode.
+ 
+You can find out what [`Gson.setPrettyPrinting`](https://www.javadoc.io/doc/com.google.code.gson/gson/2.8.5/com/google/gson/GsonBuilder.html#setPrettyPrinting--) does and how it is used in the `writeJsonFile` method of [`FileUtil.java`](https://github.com/reposense/RepoSense/blob/master/src/main/java/reposense/util/FileUtil.java).
+
+Add the following content to `FileUtil`.
+
+```
+private static boolean isPrettyPrintingUsed = false;
+```
+
+Replace the [section](https://github.com/reposense/RepoSense/blob/9125d4b6bea1e5bb6329cb8f7c476ea18fed3cea/src/main/java/reposense/util/FileUtil.java#L97-L101) in the `writeJsonFile` method with the following content.
+
+```
+GsonBuilder gsonBuilder = new GsonBuilder()
+        .setDateFormat(GITHUB_API_DATE_FORMAT)
+        .registerTypeAdapter(FileType.class, new FileType.FileTypeSerializer());
+Gson gson;
+if (isPrettyPrintingUsed) {
+    gson = gsonBuilder.setPrettyPrinting().create();
+} else {
+    gson = gsonBuilder.create();
+}
+```
+
+To notify `FileUtil` of the switch between different printing mode, add the following content to `FileUtil`.
+
+```
+public static void setPrettyPrintingMode(boolean isPrettyPrintingAdopted) {
+    isPrettyPrintingUsed = isPrettyPrintingAdopted;
+}
+```
+
+It is now possible to notify `FileUtil` of the printing mode switch by adding the following content to the `main` method of `RepoSense.java`.
+
+```
+FileUtil.setPrettyPrintingMode(cliArguments.isPrettyPrintingUsed());
+```
+
+Now the parsing of argument and changing of printing mode should have been completed.
+</panel>
+
+<panel header="Suggested solution">
+  There is more than 1 way to achieve this. One solution is shown as the following:
+  
+  According to hint 1, hint 2, and hint 3, here are the complete [required changes](https://gist.github.com/HCY123902/e85ad0f7b258937945ce1aa97fe4af9e/revisions) (See the latest revision).
```

Comment:
Similarly, try to avoid giving exact code. Unlike AB3, RepoSense code is expected to evolve all the time --> the risk of code solutions can go out of sync pretty fast.

Reply:
Sure, no problem